cgen: call fn as a map value with mut args

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
```
OS: macos, macOS, 12.3.1, 21E258
Processor: 8 cpus, 64bit, little endian, Apple M1
CC version: Apple clang version 13.1.6 (clang-1316.0.21.2.5)

getwd: /Users/hunam/Documents/dev/vweb
vmodules: /Users/hunam/.vmodules
vroot: /Applications/v
vexe: /Applications/v/v
vexe mtime: 2022-05-18 14:31:59
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 ac90a2b.5b96f7e

Git version: git version 2.32.1 (Apple Git-133)
Git vroot status: weekly.2022.20-47-ge57080f3 (4 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-unknown-unknown de82a130
```

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
type Foo = map[int]fn (mut bar Bar)

struct Bar {}

fn main() {
	foo := Foo({
		0: fn (mut bar Bar) {}
	})

	mut bar := Bar{}
	foo[0](mut bar)
}
```
 
**What did you see instead?**
```c
/tmp/v_501/report.6784010565929275243.tmp.c:2083:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/report.6784010565929275243.tmp.c:11557:100: error: passing 'main__Bar' (aka 'struct main__Bar') to parameter of incompatible type 'main__Bar *' (aka 'struct main__Bar *'); take the address with &
        ((void (*) (main__Bar* bar))(*(voidptr*)map_get(ADDR(map, foo), &(int[]){0}, &(voidptr[]){ 0 })))(bar);
                                                                                                          ^~~
                                                                                                          &
1 warning and 1 error generated.
```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -559,12 +559,12 @@ pub fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) 
 	}
 	if !found && mut node.left is ast.IndexExpr {
 		c.expr(node.left)
-		sym := c.table.sym(node.left.left_type)
+		sym := c.table.final_sym(node.left.left_type)
 		if sym.info is ast.Array {
 			elem_sym := c.table.sym(sym.info.elem_type)
 			if elem_sym.info is ast.FnType {
-				node.return_type = elem_sym.info.func.return_type
-				return elem_sym.info.func.return_type
+				func = elem_sym.info.func
+				found = true
 			} else {
 				c.error('cannot call the element of the array, it is not a function',
 					node.pos)
@@ -572,23 +572,21 @@ pub fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) 
 		} else if sym.info is ast.Map {
 			value_sym := c.table.sym(sym.info.value_type)
 			if value_sym.info is ast.FnType {
-				node.return_type = value_sym.info.func.return_type
-				return value_sym.info.func.return_type
+				func = value_sym.info.func
+				found = true
 			} else {
 				c.error('cannot call the value of the map, it is not a function', node.pos)
 			}
 		} else if sym.info is ast.ArrayFixed {
 			elem_sym := c.table.sym(sym.info.elem_type)
 			if elem_sym.info is ast.FnType {
-				node.return_type = elem_sym.info.func.return_type
-				return elem_sym.info.func.return_type
+				func = elem_sym.info.func
+				found = true
 			} else {
 				c.error('cannot call the element of the array, it is not a function',
 					node.pos)
 			}
 		}
-		found = true
-		return ast.string_type
 	}
 	if !found && mut node.left is ast.CallExpr {
 		c.expr(node.left)

--- a/vlib/v/checker/tests/fn_type_mismatch.out
+++ b/vlib/v/checker/tests/fn_type_mismatch.out
@@ -1,7 +1,20 @@
 vlib/v/checker/tests/fn_type_mismatch.vv:11:15: error: invalid array element: expected `fn (int, int) f32`, not `fn (f32, f32) f32`
-    9 | 
+    9 |
    10 | fn main() {
    11 |     fns := [add, div]
       |                  ~~~
    12 |     println(fns[0](10.0, 5.0))
    13 |     println(fns[1](10.0, 5.0))
+vlib/v/checker/tests/fn_type_mismatch.vv:12:17: error: cannot use `float literal` as `int` in argument 1 to ``
+   10 | fn main() {
+   11 |     fns := [add, div]
+   12 |     println(fns[0](10.0, 5.0))
+      |                    ~~~~
+   13 |     println(fns[1](10.0, 5.0))
+   14 | }
+vlib/v/checker/tests/fn_type_mismatch.vv:13:17: error: cannot use `float literal` as `int` in argument 1 to ``
+   11 |     fns := [add, div]
+   12 |     println(fns[0](10.0, 5.0))
+   13 |     println(fns[1](10.0, 5.0))
+      |                    ~~~~
+   14 | }

--- a/vlib/v/tests/map_get_anon_fn_value_with_mut_arg_test.v
+++ b/vlib/v/tests/map_get_anon_fn_value_with_mut_arg_test.v
@@ -1,0 +1,17 @@
+type Foo = map[int]fn (mut bar Bar) int
+
+struct Bar {}
+
+fn test_map_get_anon_fn_value_with_mut_arg() {
+	foo := Foo({
+		0: fn (mut bar Bar) int {
+			return 22
+		}
+	})
+
+	mut bar := Bar{}
+	ret := foo[0](mut bar)
+
+	println(ret)
+	assert ret == 22
+}

--- a/vlib/v/tests/map_get_anon_fn_value_with_mut_arg_test.v
+++ b/vlib/v/tests/map_get_anon_fn_value_with_mut_arg_test.v
@@ -1,13 +1,11 @@
-type Foo = map[int]fn (mut bar Bar) int
-
 struct Bar {}
 
 fn test_map_get_anon_fn_value_with_mut_arg() {
-	foo := Foo({
+	foo := {
 		0: fn (mut bar Bar) int {
 			return 22
 		}
-	})
+	}
 
 	mut bar := Bar{}
 	ret := foo[0](mut bar)

